mm2src/coins/tendermint/tendermint_coin.rs

```diff
@@ -0,0 +1,660 @@
+    account_prefix: String,
+    priv_key: Vec<u8>,
+    decimals: u8,
+    denom: Denom,
+    chain_id: ChainId,
+    sequence_lock: AsyncMutex<()>,
+}
+
+#[derive(Clone)]
+pub struct TendermintCoin(Arc<TendermintCoinImpl>);
+
+impl Deref for TendermintCoin {
+    type Target = TendermintCoinImpl;
+
+    fn deref(&self) -> &Self::Target { &self.0 }
+}
+
+pub struct TendermintInitError {
+    pub ticker: String,
+    pub kind: TendermintInitErrorKind,
+}
+
+#[derive(Display)]
+pub enum TendermintInitErrorKind {
+    InvalidPrivKey(String),
+    CouldNotGenerateAccountId(String),
+    EmptyRpcUrls,
+    RpcClientInitError(String),
+    InvalidChainId(String),
+    InvalidDenom(String),
+    RpcError(String),
+}
+
+#[derive(Display)]
+enum TendermintCoinRpcError {
+    Prost(prost::DecodeError),
+    InvalidResponse(String),
+    PerformError(String),
+}
+
+impl From<prost::DecodeError> for TendermintCoinRpcError {
+    fn from(err: DecodeError) -> Self { TendermintCoinRpcError::Prost(err) }
+}
+
+impl From<TendermintCoinRpcError> for WithdrawError {
+    fn from(err: TendermintCoinRpcError) -> Self { WithdrawError::Transport(err.to_string()) }
+}
+
+impl From<TendermintCoinRpcError> for BalanceError {
+    fn from(err: TendermintCoinRpcError) -> Self {
+        match err {
+            TendermintCoinRpcError::InvalidResponse(e) => BalanceError::InvalidResponse(e),
+            TendermintCoinRpcError::Prost(e) => BalanceError::InvalidResponse(e.to_string()),
+            TendermintCoinRpcError::PerformError(e) => BalanceError::Transport(e),
+        }
+    }
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+impl From<cosmrs::rpc::Error> for TendermintCoinRpcError {
+    fn from(err: cosmrs::rpc::Error) -> Self { TendermintCoinRpcError::PerformError(err.to_string()) }
+}
+
+#[cfg(target_arch = "wasm32")]
+impl From<PerformError> for TendermintCoinRpcError {
+    fn from(err: PerformError) -> Self { TendermintCoinRpcError::PerformError(err.to_string()) }
+}
+
+fn account_id_from_privkey(priv_key: &[u8], prefix: &str) -> MmResult<AccountId, TendermintInitErrorKind> {
+    let signing_key =
+        SigningKey::from_bytes(priv_key).map_to_mm(|e| TendermintInitErrorKind::InvalidPrivKey(e.to_string()))?;
+
+    signing_key
+        .public_key()
+        .account_id(prefix)
+        .map_to_mm(|e| TendermintInitErrorKind::CouldNotGenerateAccountId(e.to_string()))
+}
+
+fn upper_hex(bytes: &[u8]) -> String {
+    let mut str = hex::encode(bytes);
+    str.make_ascii_uppercase();
+    str
+}
+
+impl TendermintCoin {
+    pub async fn init(
+        ticker: String,
+        protocol_info: TendermintProtocolInfo,
+        activation_params: TendermintActivationParams,
+        priv_key: &[u8],
+    ) -> MmResult<Self, TendermintInitError> {
+        if activation_params.rpc_urls.is_empty() {
+            return MmError::err(TendermintInitError {
+                ticker,
+                kind: TendermintInitErrorKind::EmptyRpcUrls,
+            });
+        }
```

Comment:
A note for future refactoring, it can be nice to have a generic function that validates each payload param in `platform_coin_with_tokens` and `token` modules.